core/src/main/java/com/microsoft/applicationinsights/internal/util/ExceptionStats.java

```diff
@@ -0,0 +1,134 @@
+        this.introMessage = introMessage;
+        this.intervalSeconds = intervalSeconds;
+    }
+
+    public void recordSuccess() {
+        synchronized (lock) {
+            numSuccesses++;
+        }
+    }
+
+    // warningMessage should have low cardinality
+    public void recordFailure(String warningMessage) {
+        recordFailure(warningMessage, null);
+    }
+
+    // warningMessage should have low cardinality
+    public void recordFailure(String warningMessage, @Nullable Throwable exception) {
+        if (!firstFailure.getAndSet(true)) {
+            // log the first time we see an exception as soon as it occurs, along with full stack trace
+            logger.warn(introMessage + " " + warningMessage + " (future failures will be aggregated and logged once every " + intervalSeconds / 60 + " minutes)", exception);
+            scheduledExecutor.scheduleAtFixedRate(new ExceptionStatsLogger(), intervalSeconds, intervalSeconds, TimeUnit.SECONDS);
+            return;
+        }
+
+        logger.debug(introMessage + " " + warningMessage, exception);
+
+        if (warningMessages.size() > 100) {
+            // we have a cardinality problem and don't want to consume too much memory or do too much logging
+            return;
+        }
+        synchronized (lock) {
+            warningMessages.computeIfAbsent(warningMessage, key -> new MutableLong()).increment();
+            numFailures++;
+        }
+    }
+
+    private static class MutableLong {
+        private long value;
+        private void increment() {
+            value++;
+        }
+    }
+
+    public class ExceptionStatsLogger implements Runnable {
+
+        @Override
+        public void run() {
+            long numSuccesses;
+            long numFailures;
+            Map<String, MutableLong> warningMessages;
+            // grab quickly and reset under lock (do not perform logging under lock)
+            synchronized (lock) {
+                numSuccesses = ExceptionStats.this.numSuccesses;
+                numFailures = ExceptionStats.this.numFailures;
+                warningMessages = ExceptionStats.this.warningMessages;
+
+                ExceptionStats.this.numSuccesses = 0;
+                ExceptionStats.this.numFailures = 0;
+                ExceptionStats.this.warningMessages = new HashMap<>();
+            }
+            if (numFailures > 0) {
+                long numMinutes = ExceptionStats.this.intervalSeconds / 60;
+                long total = numSuccesses + numFailures;
+                StringBuilder message = new StringBuilder();
+                message.append("In the last ");
+                message.append(numMinutes);
+                message.append(" minutes, the following operation has failed ");
+                message.append(numFailures);
+                message.append(" times (out of ");
+                message.append(total);
+                message.append(" total):\n");
+                message.append(introMessage);
```

Comment:
Adding introMessage again is adding too much clutter I think.

Reply:
I think it's good context, but I shortened the intro significantly to make the messages less clutterful